### PR TITLE
EVEREST-178: validate s3 access

### PR DESCRIPTION
**Pull request by michal-kralik**
[EVEREST-178](https://jira.percona.com/browse/EVEREST-178) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

**CHANGE DESCRIPTION**
---
**Problem:**
EVEREST-178

*Short explanation of the problem.*
We did not validate get/put/delete permissions in an S3 bucket.

**Cause:**
*Short explanation of the root cause of the issue if applicable.*

**Solution:**
*Short explanation of the solution we are providing with this PR.*

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?

**Tests**
- [ ] Is an Integration test/test case added for the new feature/change?
- [ ] Are unit tests added where appropriate?

### CODE REVIEW

File: api/validation.go [hunk 1]
```diff
@@ -166,7 +167,33 @@ func s3Access(endpoint *string, accessKey, secretKey, bucketName, region string)
...
+	if err != nil {
+		return errors.Wrap(err, "could not read from S3 bucket")
+	}
+
+	_, err = svc.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(bucketName),
+		Key:    aws.String(testKey),
+	})
+	if err != nil {
+		return errors.Wrap(err, "could not delete an object from S3 bucket")
```
Comment (oksana-grishchenko) [inline api/validation.go:196]: Please check also the customer-faced errors for both the create and the update storage methods, looks like this error messages are gonna be overridden by a general "Could not connect to the backup storage, please check the new credentials are correct" which is confusing.

**Review by gen1us2k — COMMENTED** [review]
Can we have at least one integration test and issue created to improve this in the future when we have something s3 compatible to finish test automation

**Comment by michal-kralik** [comment]
Sure - EVEREST-453.
We already have a bunch of tests in `api-tests/tests/backup-storages.spec.ts`. We can figure out how to do S3 integration testing in future PRs.

**Comment by gen1us2k** [comment]
@michal-kralik can we have a test for this error?
```
Could not connect to the backup storage, please check the new credentials are correct
```

**Comment by michal-kralik** [comment]
> can we have a test for this error?
> 
> ```
> Could not connect to the backup storage, please check the new credentials are correct
> ```

Sure, added.

File: api/validation.go [hunk 2]
```diff
@@ -206,7 +236,11 @@ func validateCreateBackupStorageRequest(ctx echo.Context, l *zap.SugaredLogger)
 	}
 
 	// check data access
-	if err := validateStorageAccessByCreate(params); err != nil {
+	if err := validateStorageAccessByCreate(params, l); err != nil {
+		if errors.Is(err, errUserFacingMsg) {
+			return nil, errors.Wrap(err, "Could not connect to the backup storage, please check the new credentials are correct")
```
Comment (gen1us2k) [inline api/validation.go:241]: Please migrate to the recent changes and make the error in lowercase

**Review by gen1us2k — APPROVED** [review]

**Review by oksana-grishchenko — APPROVED** [review]
That's a cool idea with that special type of errors!

**Pull request merged**